odoo/tools/date_utils.py

```diff
@@ -199,6 +199,35 @@ def subtract(value, *args, **kwargs):
     """
     return value - relativedelta(*args, **kwargs)
 
+def ceil(value, granularity):
+    """
+    Return the ``value`` rounded up to a ``granularity``.
+
+    :param value: initial date or datetime.
+    :param granularity: year, month, day, hour, minute or second
+    :return: the resulting date/datetime.
+    """
+    granularities = (
+        ['year', 'month', 'day', 'hour', 'minute', 'second', 'microsecond']
+        if isinstance(value, datetime) else
+        ['year', 'month', 'day']
+    )
+    index = granularities.index(granularity, 0, len(granularities) - 1)
+    minvalues = {'month': 1, 'day': 1}
+    carry = any(
+        getattr(value, gran) != minvalues.get(gran, 0)
+        for gran in granularities[index + 1:]
+    )
+    return add(
+        value,
+        **{f'{gran}s': 1 for gran in [granularity] if carry},
+        **{
+            gran: minvalues.get(gran, 0)
+            for gran
+            in granularities[index + 1:]
+        }
```

Comment:
Just create one dict outside the `add` call procedurally. A comment on the details of `relativedelta` absolute/relative namings and how this relates to that would also be useful.

Also if you're going to use `index + 1` everywhere, just define `index` as incremented by one in the first place.